src/components/vector_path.rs

```diff
@@ -0,0 +1,89 @@
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use gtk::{glib, graphene, gsk};
+
+use tdlib::types::ClosedVectorPath;
+
+mod imp {
+    use super::*;
+    use std::cell::RefCell;
+
+    #[derive(Default)]
+    pub(crate) struct VectorPath {
+        pub(super) node: RefCell<Option<gsk::RenderNode>>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for VectorPath {
+        const NAME: &'static str = "ComponentsVectorPath";
+        type Type = super::VectorPath;
+        type ParentType = gtk::Widget;
+
+        fn class_init(klass: &mut Self::Class) {
+            klass.set_css_name("vectorpath");
+        }
+    }
+
+    impl ObjectImpl for VectorPath {}
+
+    impl WidgetImpl for VectorPath {
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            let widget = self.obj();
+
+            let factor = (widget.width() as f32).max(widget.height() as f32) / 512.0;
+            snapshot.scale(factor, factor);
+
+            if let Some(node) = &*self.node.borrow() {
+                snapshot.append_node(node);
+            }
+        }
+    }
+}
+
+glib::wrapper! {
+    pub(crate) struct VectorPath(ObjectSubclass<imp::VectorPath>)
+        @extends gtk::Widget;
+}
+
+impl VectorPath {
+    pub fn new(outline: &[ClosedVectorPath]) -> Self {
+        let obj: Self = glib::Object::new();
+        obj.imp().node.replace(Self::path_node(outline));
+        obj
+    }
+
+    fn path_node(outline: &[ClosedVectorPath]) -> Option<gsk::RenderNode> {
+        use tdlib::enums::VectorPathCommand::{CubicBezierCurve, Line};
+        use tdlib::types::VectorPathCommandCubicBezierCurve as Curve;
+
+        let snapshot = gtk::Snapshot::new();
+        let context = snapshot.append_cairo(&graphene::Rect::new(0.0, 0.0, 512.0, 512.0));
+
+        context.set_source_rgba(0.5, 0.5, 0.5, 0.4);
```

Comment:
It would be cool to get this from the widget foreground.

https://docs.gtk.org/gtk4/method.Widget.get_color.html

I guess that you need to cache it and update the color when it changes in snapshot though.

Reply:
Another option is to use https://docs.gtk.org/gtk4/iface.SymbolicPaintable.html instead of a widget.

I think you don't even need a cache anymore, because the rendering is already restrained. See https://gitlab.gnome.org/GNOME/epiphany/-/blob/master/src/ephy-downloads-paintable.c for an example.